src/extra/components/detector_motors.py

```diff
@@ -0,0 +1,282 @@
+import numpy as np
+import xarray
+
+
+def mangle_device_id_underscore(device_id):
+    return ''.join(c if c.isalnum() else '_' for c in device_id)
+
+
+def mangle_device_id_camelcase(device_id):
+    # Replace non alpha-numeric chars with blanks
+    mangled_id = ''.join(c if c.isalnum() else ' ' for c in device_id)
+    # Make camelCase
+    mangled_id = ''.join(mangled_id.title().split())
+    if mangled_id:
+        mangled_id = mangled_id[0].lower() + mangled_id[1:]
+    return mangled_id
+
+
+def check_data_selector_nodes(source, keys):
+    nodes = set((
+        key.partition('.')[0] for key in source.keys()
+    ))
+    return all(k in nodes for k in keys)
+
+
+class DetectorMotors:
+    """Interface to detector quadrant motors.
+
+    Example usage in a Jupyter notebook:
+    ```python
+            -----------------------------------------------------------
+    In [1]: |motors = DetectorMotors(run)                             |
+            |xgm                                                      |
+            -----------------------------------------------------------
+    Out[1]: <DetectorMotors SPB_IRU_AGIPD1M/MOTOR/Q{1..4}M{1..2} at
+            2023-04-04T17:44:46.844869000>
+    ```
+    """
+
+    # SPB
+    # SPB_IRU_AGIPD1M/MOTOR/Q{q+1}M{m+1}
+    # SPB_IRU_AGIPD1M/MOTOR/Z_STEPPER
+
+    # MID
+    # MID_EXP_AGIPD1M/MOTOR/Q1M1
+    # MID_EXP_UPP/MOTOR/T4
+    _position_key = "actualPosition"
+
+    DETECTORS = {
+        "SPB_IRU_AGIPD1M": {
+            "num_groups": 4,
+            "num_motors": 2,
+            "motor_pattern": "{detector_id}/MOTOR/Q{q}M{m}",
+            "data_selector": "SPB_IRU_AGIPD1M1/MDL/DATA_SELECTOR",
+        },
+        "MID_EXP_AGIPD1M": {
+            "num_groups": 4,
+            "num_motors": 2,
+            "motor_pattern": "{detector_id}/MOTOR/Q{q}M{m}",
+            "data_selector": "MID_EXP_AGIPD1M1/MDL/DATA_SELECTOR",
+        },
+    }
+    ARGS = ("num_groups", "num_motors", "motor_pattern", "data_selector")
+
+    def __init__(self, dc, detector_id="", groups=None, motors_per_group=None,
```

Comment:
If only `dc` is passed, do we want to check for the detector IDs we know about, and if exactly one of them is found, use that? I think this is a common pattern for components, that `Component(run)` tries to find the relevant sources if it can.

(This is also a point in favour of making detector-specific subclasses, because it lets you be more specific without having to copy the device ID)

Reply:
I've generalized the base class a bit and created child class for AGIPD1M quadrants. In the last one I have implemented the search of the any suitable motors as sources or within any `DataSelector`. The data access logic remains in the base class.

Is it better now?